docs/proposals/20240327-enhancedlivenessprobe.md

```diff
@@ -0,0 +1,189 @@
+Secondly, there is no standard concept in the design idea of community method, 
+which completely relies on the machine node detection and machine node restart mechanism. 
+Furthermore, for a stability perspective, community-native solutions lack application-level global high availability and any resilience protection policies. 
+Combined with the existing defects, it proposes to design and implement the enhanced livenessProbe application survival solution in the OpenKruise suite, 
+and plans to open source release as the core capability in the future, so that the innovative technology can be used by more people.
+
+## Proposal
+
+The enhanced livenessProbe solution are composed of two controllers、one webhook and a node probe detection component, 
+there are an enhanced livenessProbe webhook, an enhanced livenessProbe map nodePodProbe resource controller, 
+an enhanced livenessProbe controller and a node detection component.
+
+● **EnhancedLivenessProbeWebhook**: A webhook works in Kruise-manager. 
+The native livenessProbe is converted to the specific string configuration field in pod annotations as soon as the pod is created.
+
+● **EnhancedLivenessProbeMapNodePodProbeController**: It is a service component controller which is developed in Kruise-manager.
+The specific string configuration known as the livenessProbe is converted to the nodePodProbe(a custom resource in Kruise suite) resource immediately.
+
+● **EnhancedLivenessProbeController**: It is a service component controller which is developed in Kruise-manager.
+It watches the state change of the nodePodProbe resource and analyzes the application workload status with the failed livenessProbe pods. 
+This controller decide whether to restart containers or not(using the container recreate request, a custom resource in Kruise). 
+In calculating process, the resilience protection and application maxUnavailable protection policies are considered by the special algorithm and detection method comprehensively.
+
+● **LivenessProbeDetector**: It is just a single controller in Kruise-daemon, 
+which support to check the nodePodProbe(e.g., tcp/http/exec checking) as the same by the kubelet probe logic, using the node pod probe resource in Kruise.
+
+### Implementation
+
+**EnhancedLivenessProbeWebhook** 
+> When a pod created with livenessProbe config with `Annotations: apps.kruise.io/using-enhanced-liveness: true` feature gate, 
+the webhook will admit the pod resource to remove the native livenessProbe and patch the configuration to the pod annotations `apps.kruise.io/container-probe-context: '{...json format for liveness probe with specified contaienrs in pod...}''`.
+
+**EnhancedLivenessProbeMapNodePodProbeController**
+> For using the nodPodProbe controller processing, 
+> the livenessProbe config filed in the pod annotations should be converted to the nodePodProbe custom defined resource in Kruise suite. 
+> This controller can create and update the nodePodProbe resource.
```

Comment:
plz explain in detail about the reconcile logic in pod creating, deleting , exited case